bid-distribution/src/settlement_claims.rs

```diff
@@ -0,0 +1,169 @@
+use log::info;
+use rust_decimal::prelude::*;
+use rust_decimal::Decimal;
+use settlement_engine::settlement_claims::{
+    Settlement, SettlementClaim, SettlementCollection, SettlementReason,
+};
+use settlement_engine::stake_meta_index::StakeMetaIndex;
+use solana_sdk::pubkey::Pubkey;
+use std::cmp::min;
+use std::collections::HashMap;
+
+pub fn generate_bid_settlement_collection(
+    stake_meta_index: &StakeMetaIndex,
+    sam_validator_metas: &Vec<ValidatorSamMeta>,
+    stake_authority_filter: &dyn Fn(&Pubkey) -> bool,
+    settlement_config: &SettlementConfig,
+) -> SettlementCollection {
+    let settlements = generate_bid_settlements(
+        stake_meta_index,
+        sam_validator_metas,
+        &stake_authority_filter,
+        settlement_config,
+    );
+
+    SettlementCollection {
+        slot: stake_meta_index.stake_meta_collection.slot,
+        epoch: stake_meta_index.stake_meta_collection.epoch,
+        settlements,
+    }
+}
+
+pub fn generate_bid_settlements(
+    stake_meta_index: &StakeMetaIndex,
+    sam_validator_metas: &Vec<ValidatorSamMeta>,
+    stake_authority_filter: &dyn Fn(&Pubkey) -> bool,
+    settlement_config: &SettlementConfig,
+) -> Vec<Settlement> {
+    info!("Generating bid settlements...");
+
+    assert!(
+        sam_validator_metas
+            .iter()
+            .all(|v| v.epoch as u64 == stake_meta_index.stake_meta_collection.epoch),
+        "SAM Validators Collection epoch must be same as stake meta collection epoch"
+    );
+
+    let marinade_fee_deposit_stake_accounts: HashMap<_, _> = stake_meta_index
+        .stake_meta_collection
+        .stake_metas
+        .iter()
+        .find(|x| {
+            x.withdraw_authority
+                .eq(settlement_config.marinade_withdraw_authority())
+                && x.stake_authority
+                    .eq(settlement_config.marinade_stake_authority())
+        })
+        .iter()
+        .map(|s| (s.pubkey, s.active_delegation_lamports))
+        .collect();
+
+    let mut settlement_claim_collections = vec![];
+
+    for validator in sam_validator_metas {
+        if let Some(grouped_stake_metas) =
+            stake_meta_index.iter_grouped_stake_metas(&validator.vote_account)
+        {
+            if validator.effective_bid == Decimal::ZERO {
+                continue;
+            }
+            let sam_target_stake =
+                validator.marinade_sam_target_sol * Decimal::from_f64(1e9).unwrap();
+            let mnde_target_stake =
+                validator.marinade_mnde_target_sol * Decimal::from_f64(1e9).unwrap();
+            let max_wanted_stake = validator.max_stake_wanted * Decimal::from_f64(1e9).unwrap();
+            let marinade_payment_percentage =
+                Decimal::from(*settlement_config.marinade_fee_bps()) / Decimal::from(10000);
+            let effective_bid = validator.effective_bid / Decimal::from(1000);
+
+            if sam_target_stake + mnde_target_stake == Decimal::ZERO {
```

Comment:
this is not correct @butonium @ochaloup 

case:
- validator A already had 1 M SOL stake from Marinade
- for whatever reason, the next auction run shows the target for the validator is 0
- we still charge bid in this case!

Reply:
https://github.com/marinade-finance/ds-sam/blob/main/src/commands/analyze-revenue.cmd.ts#L145